src/object_based/htm.cpp

```diff
@@ -0,0 +1,142 @@
+#include "htm.hpp"
+
+namespace ear {
+  using Point = HTM::Point;
+  using ConstPointRef = const Eigen::Ref<const Eigen::Vector3d> &;
+  using Tri = Eigen::Matrix3d;
+
+  namespace {
+    /// normalised mid-point between two points
+    Point midpoint(ConstPointRef p1, ConstPointRef p2) {
+      Point p = p1 + p2;
+      return p.normalized();
+    }
+
+    /// sub-divide a triangle into 4
+    /// NOTE: add_tris depends on the exact order of triangles and corners
+    /// returned
+    std::vector<Tri> next_tris(ConstPointRef p1, ConstPointRef p2,
+                               ConstPointRef p3) {
+      std::vector<Tri> tris;
+
+      Point m12 = midpoint(p1, p2);
+      Point m23 = midpoint(p2, p3);
+      Point m31 = midpoint(p3, p1);
+
+      Tri t;
+
+      t << p1, m12, m31;
+      tris.push_back(t);
+
+      t << p2, m23, m12;
+      tris.push_back(t);
+
+      t << p3, m31, m23;
+      tris.push_back(t);
+
+      t << m12, m23, m31;
+      tris.push_back(t);
+
+      return tris;
+    }
+
+    /// get the 8 initial triangles
+    std::vector<Tri> initial_tris() {
```

Comment:
Could this be array<Tri, 8> as well? Almost certainly nit picking in this context but I think std::arrays go on the stack & don't waste space (whereas a vector without a .reserve() can grab a bunch of heap memory) so try to default to them where I can out of some sort of possibly misguided principle.